File: 3_Regression/Multi_Linear_Regression/R/multiple_linear_regression.R
# Multiple Linear Regression

# Importing the dataset
dataset = read.csv('50_Startups.csv')

# Encoding categorical data
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Multiple Linear Regression to the Training set
regressor = lm(formula = Profit ~ .,
               data = training_set)

# Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)

# Building the optimal model using Backward Elimination
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)
# Optional Step: Remove State2 only (as opposed to removing State directly)
# regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + factor(State, exclude = 2),
#                data = dataset)
# summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)

# Regression subset selection including exhaustive search
library(leaps)
#Perform all subset regression, and choose “nbest” model(s) for each number of predictors
leaps<-regsubsets(Profit ~ .,data=training_set, nbest=1, method = "exhaustive")

# view results and best model at each variable number
summary(leaps)

# Graphical table of best subsets, models are ordered by the selection statistic
plot(leaps,scale="r2")

# plot statistic by subset size
subsets(leaps, statistic="adjr2", legend = FALSE)

# Mallow Cp is used to decide on the number of predictors to include
subsets(leaps, statistic="cp", legend = FALSE)

File: 3_Regression/Multi_Linear_Regression/Python/multiple_linear_regression.py
""" Multiple Linear Regression
"""

# Importing the libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import statsmodels.formula.api as sm


def main():
    # Importing the dataset
    dataset = pd.read_csv('../50_Startups.csv')
    X = dataset.iloc[:, :-1].values
    y = dataset.iloc[:, 4].values

    # Encoding categorical data
    labelencoder = LabelEncoder()
    X[:, 3] = labelencoder.fit_transform(X[:, 3])
    onehotencoder = OneHotEncoder(categorical_features=[3])
    X = onehotencoder.fit_transform(X).toarray()

    # Avoiding the Dummy Variable Trap
    X = X[:, 1:]

    # Splitting the dataset into the Training set and Test set
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

    # Feature Scaling
    """from sklearn.preprocessing import StandardScaler
    sc_X = StandardScaler()
    X_train = sc_X.fit_transform(X_train)
    X_test = sc_X.transform(X_test)
    sc_y = StandardScaler()
    y_train = sc_y.fit_transform(y_train)"""

    # Fitting Multiple Linear Regression to the Training set
    regressor = LinearRegression()
    regressor.fit(X_train, y_train)

    # Predicting the Test set results
    y_pred = regressor.predict(X_test)

    # Building the optimal model using Backward Elimination
    X = np.append(arr=np.ones((50, 1)).astype(int), values=X, axis=1)
    X_opt = X[:, [0, 1, 2, 3, 4, 5]]
    regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
    regressor_OLS.summary()
    X_opt = X[:, [0, 1, 3, 4, 5]]
    regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
    regressor_OLS.summary()
    X_opt = X[:, [0, 3, 4, 5]]
    regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
    regressor_OLS.summary()
    X_opt = X[:, [0, 3, 5]]
    regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
    regressor_OLS.summary()
    X_opt = X[:, [0, 3]]
    regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
    regressor_OLS.summary()


if __name__ == '__main__':
    main()
